apps/web/components/dashboard-components/project-details/custom-llm-parameters/editor.tsx

```diff
@@ -0,0 +1,352 @@
+  ]);
+
+  const updateProject = api.project.updateProject.useMutation({
+    onSuccess: () => {
+      setIsEditing(false);
+      setActiveEditIndex(null);
+      toast({
+        variant: "success",
+        title: "Success",
+        description: "Custom parameters saved successfully",
+      });
+      if (onEdited) {
+        onEdited();
+      }
+    },
+    onError: (error) => {
+      toast({
+        variant: "destructive",
+        title: "Error saving parameters",
+        description: error.message,
+      });
+    },
+  });
+
+  const hasValidationErrors = useMemo(() => {
+    return parameters
+      .filter((p) => p.key.trim())
+      .some((p) => {
+        const convertedValue = convertValue(p.value, p.type);
+        return convertedValue === undefined && p.value.trim();
+      });
+  }, [parameters]);
+
+  const handleSave = useCallback(() => {
+    if (!project || !currentProvider || !currentModel) return;
+
+    // Convert UI parameters back to their proper types
+    const parametersObject = Object.fromEntries(
+      parameters
+        .filter((p) => p.key)
+        .map((p) => [p.key, convertValue(p.value, p.type)])
+        .filter(([_, v]) => v !== undefined),
+    );
+
+    const existingParams = project.customLlmParameters ?? {};
+    const hasNewParams = Object.keys(parametersObject).length > 0;
+
+    // Get existing parameters for current provider
+    const currentProviderParams = existingParams[currentProvider] ?? {};
+
+    // Extract other models (excluding current model)
+    const { [currentModel]: _, ...otherModels } = currentProviderParams;
+    const hasOtherModels = Object.keys(otherModels).length > 0;
+
+    // Determine updated parameters for this provider
+    let updatedProviderParams;
+    if (hasNewParams) {
+      // Case A: User added new parameters
+      updatedProviderParams = {
+        ...currentProviderParams, // Keep existing model params
+        [currentModel]: parametersObject, // Add/update current model
+      };
+    } else if (hasOtherModels) {
+      // Case B: No new params, but other models exist for this provider
+      updatedProviderParams = otherModels; // Keep other models
+    } else {
+      // Case C: No params and no other models
+      updatedProviderParams = null; // Remove provider entirely
+    }
+
+    // Extract other providers (excluding current provider)
+    const { [currentProvider]: __, ...otherProviders } = existingParams;
+    const hasOtherProviders = Object.keys(otherProviders).length > 0;
+
+    let customLlmParameters;
+    if (updatedProviderParams) {
+      // Provider still has params - merge back
+      customLlmParameters = {
+        ...existingParams,
+        [currentProvider]: updatedProviderParams,
+      };
+    } else if (hasOtherProviders) {
+      // No params for current provider, but other providers exist
+      customLlmParameters = otherProviders;
+    } else {
+      // No params anywhere
+      customLlmParameters = null;
+    }
```

Comment:
This all seems like a lot of extra code, can't you just do this:

```suggestion
const customLlmParameters = {
   ...existingParams,
   [provider]: {
     ...existingParams[provider],
     [model]: parametersObject, 
    },
}
```

the *one* drawback is if you eliminate all the parameters, you're leaving behind `{}` in the database, but that seems totally reasonable.